### PR TITLE
STUD-2280 Containerd

### PR DESCRIPTION
**Pull request by austinarbor-wk**
## Purpose

Adds ability to launch command containers on a containerd backend instead of docker

## Solution

Added a new implementation for containerd that mimics the existing docker implementation

## Semantic Versioning (check one)

- [ ] The following were changed in a non-backward compatible way and requires a major version bump:
  - _[link to the breaking change in the diff]_
- [x] Something public was added or changed in backward compatible way, this requires a minor version bump
- [ ] No public changes nor new features (backwards-compatible refactor or bug fix), so this can be included in a patch
      release

## How to QA

- [ ] run the following related examples: **(fill in)**
- [ ] Other necessary steps needed to fully exercise the solution should be added here. **(fill in)**

### CODE REVIEW

**Comment by aviary3-wk** [comment]
## Security Insights
No security relevant content was detected by automated scans.

## Action Items
* Review PR for [security impact](https://wiki.atl.workiva.net/display/SECURITY/Development+Security+Review+Guidelines); comment "security review required" if needed or unsure
* Verify [`aviary.yaml`](https://wiki.atl.workiva.net/pages/viewpage.action?pageId=42762752#SecurityInsightFramework&Tooling-ConfiguringAviary) coverage of security relevant code
---
_Questions or Comments? Reach out on Slack: #support-infosec._

**Comment by austinarbor-wk** [comment]
@danielnovotny-wf this PR has the bulk of the changes required to start using containerd. Our use case is probably slightly different than others in that our docker commands were fitted to implement an interface that matches osexec.Cmd (using this repo as our own forked version of the `go-dexec` library) so I kept that logic in place in order to minimize changes requred in the consuming application

File: id.go [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+	"time"
+)
+
+func init() {
+	rand.Seed(time.Now().UnixNano())
+}
+
+const letters = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ"
+
+func RandomString(n int) string {
```
Comment (ryancurtin-wk) [inline id.go:14]: Any reason to roll our own here?

File: execution.go [hunk 2]
```diff
@@ -1,133 +1,21 @@
...
+}
 
 // Execution determines how the command is going to be executed. Currently
 // the only method is ByCreatingContainer.
-type Execution interface {
-	create(d Docker, cmd []string) error
-	run(d Docker, stdin io.Reader, stdout, stderr io.Writer) error
-	wait(d Docker) (int, error)
+type Execution[T ContainerClient] interface {
+	create(d T, cmd []string) error
```
Comment (ryancurtin-wk) [inline execution.go:12]: This predates you and the whole thing is one package, but un-exported functions on an interface?  I guess I thought I had seen it all

**Review by ryancurtin-wk — APPROVED** [review]
LGTM, only minor comments.  The repos that consume it specify a version, so it is safe to assume there won't be any negative consequences.  Happy to handle the real QA via the PR in OneCloudInc/droplet.

File: execution.go (hunk 2 above)
Comment (austinarbor-wk) [reply execution.go:12]: yea, pretty odd but what can ya do

**Comment by austinarbor-wk** [comment]
QA +1

**Comment by austinarbor-wk** [comment]
@Workiva/release-management-p

**Review by rmconsole-wf — APPROVED** [review]
+1 from RM

**Pull request merged**